Flatten interaction

kangalio opened the pull request:
Supersedes #2039. Marked draft because collectors still need to be updated (but only once #2240 lands)

Like #2039, this PR flattens the Interaction enum into a struct to not duplicate shared data. Type specific data is stored in an enum.

Before:
```rust
enum Interaction {
    Ping(PingInteraction),
    Command(CommandInteraction),
    Autocomplete(CommandInteraction),
    Component(ComponentInteraction),
    Modal(ModalInteraction),
}

struct PingInteraction {
    /* shared fields */
}

struct CommandInteraction {
    /* shared fields */
    data: CommandData,
}

struct ComponentInteraction {
    /* shared fields */
    data: ComponentData,
}

struct ModalInteraction {
    /* shared fields */
    data: ModalData,
}
```

After:
```rust
struct Interaction {
    /* shared fields */
    kind: InteractionKind,
}

enum InteractionKind {
    Ping,
    Command(CommandData),
    Autocomplete(CommandData),
    Component(ComponentData),
    Modal(ModalData),
}
```

nickelc committed:
Remove deprecated `audit_log` re-exports and types (#1864)

The audit log types are no longer re-exported in the `model::guild` module.

BREAKING CHANGES:
- The `audit_log::Action*` enum sub types are renamed to `*Action`.
  For instance, `ActionChannel` to `ChannelAction`.

- The audit log types must be imported directly from the
  `model::guild::audit_log` module.

GnomedDev committed:
Reference the objects within `ResolvedTarget` instead of cloning them (#1885)

nickelc committed:
Replace `gateway::WsStream` & extension traits with a newtype struct (#1893)

The newtype struct `gateway::WsClient` hides the WebSocket
implementation details and is the first step to remove the
`async_tungstenite` & `tungstenite` types from the public API.

The `WebSocketGatewayClientExt`, `ReceiverExt` and `SenderExt` trait
methods are merged into the implementation of the newtype struct.

BREAKING CHANGES:
- `gateway::WsStream` is replaced by `gateway::WsClient`.
- The `gateway::WebSocketGatewayClientExt` trait is removed.

Miezhiko committed:
Change `Typing::stop` to return a `bool` (#1897)

This changes the return type of `http::Typing::stop` to be a simple
`bool` rather than `Option<()>`, which doesn't provide useful
information that couldn't be expressed with just a `bool`.

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>
Signed-off-by: Miezhiko <Miezhiko@gmail.com>

nickelc committed:
Remove useless reference on `Option<&T>` (#1896)

Since `&` is Copy, it’s useless to have a reference on `Option<&'a [u8]>`.

nickelc committed:
Remove deprecated interaction type aliases, field and method (#1902)

GnomedDev committed:
Switch to `Into<String>` when accepting string parameters (#1903)

This changes instances of `ToString` to `Into<String>` whenever ownership of
the `String` is necessary. In case it is not, it's simply changed to `&str` to
lessen the effects of monomorphization on compile times.

vaporoxx committed:
Allow having no framework when the `framework` feature is enabled (#1908)

vaporoxx committed:
Do not set message content not sent by the API (#1910)

Problems with the current setup:

- The content of messages with these types is not documented, meaning that it can change anytime.
- Not every message type is handled, e.g. boost or thread creation messages do not get their content added.
- Overall more possibly unexpected behaviour and less consistency with the API.

GnomedDev committed:
Remove unneeded `futures::iter::stream` usage (#1922)

mkrasnitski committed:
Add support for executing webhooks in threads (#1923)

This adds a method `Webhook::execute_in_thread` which takes an additional
thread ID (which is just a `ChannelId`) parameter. Then, the webhook will
execute in that thread, provided it belongs to the webhook's assigned guild
channel.

This also adds a `thread_id` parameter to `Http::execute_webhook`, which is a
breaking change.

GnomedDev committed:
Add a cache fallback to `ChannelId::message` (#1924)

GnomedDev committed:
Change `User::has_role` to accept just a `GuildId` (#1927)

This also removes `GuildContainer`, as this was (and historically has been) the only use of it.

GnomedDev committed:
Implement suggested `TODO` fixes (#1928)

GnomedDev committed:
Expose `dashmap`'s wrapped `Ref` to reduce cloning (#1917)

GnomedDev committed:
Move around `Box`es to fix `clippy::large_enum_variant` (#1929)

GnomedDev committed:
Fix clippy warnings (#1933)

GnomedDev committed:
Avoid cloning member out of cache in `Guild::member` (#1934)

GnomedDev committed:
De-duplicate cache -> http logic from `Message::member` (#1935)

mkrasnitski committed:
Clean up previous changes to cache (#1936)

This replaces the implementation of `From<Ref>` for `CacheRef` with a
private method in order to not expose the dashmap type as part of the
public API.

Also, this rewords and removes a few clones on types returned from the cache.

vaporoxx committed:
Remove references from `ShardManagerOptions` (#1937)

nickelc committed:
Unasync `ShardManager::new` (#1940)

nickelc committed:
Introduce a separate data type for autocomplete interactions (#1941)

The interaction payload for autocompletes is slightly different. It's
missing the `resolved` data map and reports only for the current
autocompleted option value the `focused` field and its value as a string.

The focused option value is modeled as an enum variant of `CommandDataOptionValue`.

```
enum CommandDataOptionValue {
    Autocomplete { name: String, kind: CommandOptionType, value: String },
}
```

The focused option value can be retrieved via `AutocompleteData::focused_option()`
from the `AutocompleteData::options` vector.

GnomedDev committed:
Reduce int size of `prune_count` methods (#1947)

GnomedDev committed:
Clean and optimise `message_builder.rs` (#1946)

GnomedDev committed:
Remove unused allowed clippy warnings or fix them (#1945)

GnomedDev committed:
Use concrete fields instead of `HashMap` in the `GetMessages` builder (#1942)

GnomedDev committed:
Remove user-only `EditProfile` methods (#1959)

BREAKING CHANGE: This removes methods only usable on user accounts.

GnomedDev committed:
Improve event deserialization by sacrificing `GuildUnavailable` & `Unknown` events (#1938, #1951)

BREAKING CHANGES: Removes the `guild_unavailable` & `unknown` methods
from the `EventHandler` trait.

GnomedDev committed:
Rework manual deserialization implementations (#1938, #1951)

GnomedDev committed:
Fix `needless_pass_by_value` caused by new `ToString` usage (#1961)

GnomedDev committed:
Store `Vec<u8>` instead of `&[u8]` in `Request` (#1963)

GnomedDev committed:
Remove `kind` from the interaction types (#1970)

nickelc committed:
Remove unneeded `#[repr(u8)]` attributes (#1972)

GnomedDev committed:
Accept `Into<String>` in more places (#1955)

nickelc committed:
Add missing `member` field to `MessageInteraction` (#1976)

nickelc committed:
Rename `Permissions::USE_SLASH_COMMANDS` to `USE_APPLICATION_COMMANDS` (#1977)

The permission was renamed by Discord in regards to the new context menu commands.

GnomedDev committed:
Rework builder module to get rid of all `HashMap<'static str, T>`  (#1962)

vaporoxx committed:
Update `OpCode` to match Discord's documentation (#1980)

GnomedDev committed:
Add missing `#[serde(rename = "type")]` attributes (#1987)

GnomedDev committed:
Remove usage of the `json!` macro from websocket message sending (#1981)

GnomedDev committed:
Use `ReactionType` to serialize the emoji in `CreateComponents` (#1989)

GnomedDev committed:
Reuse `edit_member` in `GuildId` methods (#1991)

GnomedDev committed:
Use custom enum for `Message::nonce` instead of `json::Value` (#1990)

GnomedDev committed:
Add proper builders for `Webhook` (#1993)

This adds `CreateWebhook` and `EditWebhook`, and gets rid of the specific
methods for different combinations of arguments. This also gets rid of the
`json!` usage and `Value` serialization, instead using `derive(Serialize)`

GnomedDev committed:
Swap all ID types to use `NonZeroU64` internally (#1930)

GnomedDev committed:
Use `CacheRef` for `CurrentUser` (#1996)

This removes `Cache::current_user_id` and `current_user_field` and makes
`Cache::current_user` return a `CacheRef` by reworking the internals of
`CacheRef` to hold an enum of different guard types, currently only a dashmap
`Ref` or a `parking_lot::RwLockReadGuard`. This prevents unnecessary cloning of
the current user and monomorphization bloat from the `_field` method.

GnomedDev committed:
Store messages in a `HashMap` instead of a `DashMap` (#1995)

The message cache is currently stored as `DashMap<ChannelId, DashMap<MessageId,
Message>>` however since the outer `DashMap` already provides interior
mutability, it doesn't need another `DashMap` inside.

This also stops dashmap being exposed to the public API via
`MessageIterator::Item` by removing it (and instead exposes the inner
`HashMap`).

GnomedDev committed:
Deduplicate a lot of the collector module (#1932)

nickelc committed:
Group & sort derive macros alphabetically

The derive macros are grouped into `std` & third party and
sorted alphabetically.

```
struct Foo;
```

nickelc committed:
Unify `derive` & `#[non_exhaustive]` attributes for number enums (#1999)

BREAKING CHANGE: Some enums were missing the `[non_exhaustive]` attribute.

nickelc committed:
Remove unused `channel/sticker.rs` file

The file was forgotten to delete when the types were moved into the
root. See 34bc5fec302c1f66cc851d6626f04a8831c4f8b2

nickelc committed:
Replace magic number with `ChannelType::Stage`

nickelc committed:
Group & sort derive macros alphabetically

The derive macros are grouped into `std` & third party and
sorted alphabetically.

```
struct Foo;
```

GnomedDev committed:
Use `CacheRef` to access channel messages from the cache (#2000)

BREAKING CHANGE: `Cache::channel_messages_field` is replaced by `Cache::channel_messages`
returning a reference to the messages map.

nickelc committed:
Make `MessageActivityKind`'s variants upper camel case (#2005)

GnomedDev committed:
Reduce size of shard info and turn it into a struct (#1984)

This replaces the `[u64; 2]` shard info with `ShardInfo { id: u32, total: u32 }`.

Co-authored-by: nickelc <constantin.nickel@gmail.com>

GnomedDev committed:
Add `UserRef` by wrapping temporary cache values in `Arc` (#1997)

vaporoxx committed:
Rework setting and storing of presence data (#2007)

nickelc committed:
Preserve values for unknown enum variants (#2008)

The `enum_number!` macro now takes the whole enum definition and generates
`From` trait implementations to convert a value to the enum and back.

The implementations can then be picked up by `serde` with
`#[serde(from = "u8", into = "u8")]` to (de)serialize the types.

```
enum_number! {
    #[derive(Clone, Copy, Debug, Deserialize, Serialize)]
    #[serde(from = "u8", into = "u8")]
    pub enum Foo {
        A = 1,
        B = 2,
        _ => Unknown(u8),
    }
}
```

BREAKING CHANGE: The `Unknown` variant now takes the unknown value
and the removed `fn num() -> u64` method can be replaced
with `let v = u8::from(kind)`.

nickelc committed:
fixup! Use `CacheRef` for `CurrentUser` (#1996)

GnomedDev committed:
Remove `Shard` prefix in `ShardInfo` logging (#2014)

Before `ShardInfo`:
`[serenity::gateway::shard]: [Shard [15, 33]] ...`

Currently:
`[serenity::gateway::shard]: [Shard ShardInfo { id: 15, total: 33 }] ...`

After:
`[serenity::gateway::shard]: [ShardInfo { id: 15, total: 33 }] ...`

nickelc committed:
Remove unneeded `#[must_use]` attribute

The returned `Result` is already marked as `#[must_use]`.

GnomedDev committed:
Clean up usage of `ref`, `ref mut`, and `as_ref()` (#2016)

GnomedDev committed:
Change `Member::edit` to edit the member in-place (#2017)

GnomedDev committed:
Fix possible deadlock in `PartialGuild::greater_member_hierarchy` (#2015)

GnomedDev committed:
Remove `InterMessage::Json` (#2009)

vaporoxx committed:
Support setting presence when identifying (#2018)

GnomedDev committed:
Clean up collector module by replacing duplication with generics (#2011)

nickelc committed:
Use `GenericId` for the `AuditLogEntry::target_id` field (#2023)

GnomedDev committed:
Fix return type for the guild delete endpoint (#2030)

nickelc committed:
Replace `json::from_number` with `into()` (#2034)

Both `Value` types implement `From` for their number variants.

nickelc committed:
Merge application command & autocomplete interactions (#2031)

The structs are almost identical except for the `focused` field for the
current autocompleted option and that the `resolved` data map is not
sent for autocomplete interactions.

The entities from the `CommandData::resolved` data map are no longer
cloned for the option values during deserialization. But the resolved
options and values can be accessed via the `CommandData::options()` method.

For autocomplete interactions `CommandData::autocomplete()` returns the
current autocompleted option + value.

BREAKING CHANGES:
- `ApplicationCommandInteraction` is now used for application commands
  and their autocomplete interaction.

- The `CommandDataOptionValue`'s variants no longer have the cloned
  entities from the `CommandData::resolved` data map.
  See the `CommandData::options` method.

- The fields `value`, `kind`, `options` and `focused` of `CommandDataOption`
  are streamlined into the `CommandDataOptionValue` enum with new variants
  for subcommands, subcommand groups and autocomplete.

nickelc committed:
Switch to `tokio-tungstenite` (#2036)

GnomedDev committed:
Use `try_from` and `into` serde attributes to deserialize `PermissionOverwrite` (#2042)

GnomedDev committed:
Fix `Timestamp` having a fallible `From` impl (#2032)

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

GnomedDev committed:
Constify some functions that could be used in a `const` context (#2047)

GnomedDev committed:
Fix new clippy warnings (#2041)

arqunis committed:
Fix deserialisation for reaction events

To simplify the deserialisation implementation for `Event`, it was
changed to use serde's native support for `enum` representations that
are adjacently tagged (meaning in a text form like JSON, the tag is in
one field, the data in the other field).

This helped in trimming the lines of code we have to maintain; however,
it did break reaction events whose names are prepended `MESSAGE_`, which
serde could not account for when renaming, for instance, `ReactionAdd`
to SCREAMING_SNAKE_CASE -> `REACTION_ADD`, when it should have been
`MESSAGE_REACTION_ADD`.

This commit manually renames the reaction events to have `MESSAGE_` at
the beginning of their event names.

GnomedDev committed:
Clean up dispatch and unnecessary clones in client (#2019)

GnomedDev committed:
Make standard framework configurable at runtime (#1950)

GnomedDev committed:
Deduplicate `Filter` in collector (#2040)

arqunis committed:
Revert "Clean up dispatch and unnecessary clones in client (#2019)"

This reverts commit de05e5bb3a99b45010b5aa2ebaa64a32db2d2e97 because it
broke shard code to the point where it just crashes, preventing from
even a simple ping-pong bot from running.

Miezhiko committed:
Increase integer range of `GuildChannel::message_count` from `u8` to `u32` (#2054)

GnomedDev committed:
Deserialize to `JsonMap` then recreate `Value` when needed (#2057)

GnomedDev committed:
Use `NonZeroU32` in `CommonFilterOptions` (#2052)

arqunis committed:
Fix compilation errors that arose from rebasing

Additionally, adapt some code to the `next` branch.

GnomedDev committed:
Fix clippy warnings and apply `next` patches to new code (#2061)

Co-authored-by: Michael Krasnitski <42564254+mkrasnitski@users.noreply.github.com>

mkrasnitski committed:
Overhaul usage of the builder pattern (#2024)

GnomedDev committed:
Remove unnecessary generics (#2065)

mkrasnitski committed:
Fix some builders returning unsendable Futures (#2069)

Due to how the permissions lookups were being performed, the Futures
were being marked `!Send` because `GuildRef` is also `!Send`. So, we
unfortunately have to clone the guild we get from the cache.

GnomedDev committed:
Move `Colour` to `models` and remove feature guards (#2073)

Bloectasy committed:
Fix compilation of `Member::kick_with_reason` caused by `CacheRef` (#2072)

GnomedDev committed:
Reduce unknown event log to debug level (#2064)

mkrasnitski committed:
Remove unneeded and deprecated application commands builders. (#2078)

GnomedDev committed:
Avoid using `serde_json` in the HTTP client when `simd_json` is enabled (#2083)

mkrasnitski committed:
Remove `CreateSelectMenuOptions` builder (#2086)

GnomedDev committed:
Swap to `FxHash` instead of default hasher (#2084)

GnomedDev committed:
Remove `async` from methods that can just be synchronous (#2090)

mkrasnitski committed:
Enforce required fields for builders (#2087)

This commit adds enforcement for unconditionally required fields on builders.
It does so by removing the implementation of `Default` on those builders and
adding a `new` method that takes the required fields as argument.

mkrasnitski committed:
Fix some documentation comments and typos (#2095)

arqunis committed:
Fix issues from rebase

mkrasnitski committed:
Use fallback endpoints in `Webhook::{edit,delete}` if no token is set (#2104)

If the `Webhook::token` field is `None` when these functions are called,
attempt to use the alternate authenticated endpoint instead of erroring.
Also, add audit_log_reason parameter to `Http::edit_webhook_with_token`.

mkrasnitski committed:
Remove needless borrows of `impl AsRef<Http>` and `impl CacheHttp` (#2108)

mkrasnitski committed:
Change some model methods to take `&self` instead of `self` (#2106)

mkrasnitski committed:
Add missing audit log reason support to `Http` and relevant builders (#2105)

GnomedDev committed:
Remove `Arc` around `CacheAndHttp` (#2094)

kangalio committed:
Make `AutocompleteChoice` constructable (#2109)

oSumAtrIX committed:
Fix incorrect documentation for `GuildChannel::create_permission`

mkrasnitski committed:
Elide lifetimes on model methods (#2111)

mkrasnitski committed:
Make `description` field optional when creating slash commands (#2110)

kangalio committed:
Remove `Channel::Category` (#2128)

kangalio committed:
Alter `ChannelId::name` to fetch from HTTP when necessary (#2139)

kangalio committed:
Change channel/role position and bitrate type to `u32` (#2149)

kangalio committed:
Stop changing `EventHandler` signatures depending on the `cache` feature (#2144)

kangalio committed:
Stop considering guild emojis as mentions (#2161)

kangalio committed:
Add `EventHandler::shards_ready` (#2164)

kangalio committed:
Remove `parse_token` (#2162)

It was broken due to tokens not being an officially documented format, and the
format having changed silently since being implemented. `validate_token` still
exists, but it does less intrusive check.

kangalio committed:
Make `CreateButton` and `Button` more type-safe (#2166)

mkrasnitski committed:
Unify length overflow checks for builders (#2178)

kangalio committed:
Simplify API of `CreateAllowedMentions` (#2127)

kangalio committed:
Fix `http` and `gateway` features (#2172)

Previously, you couldn't even build Serenity if you had just `http` or just `gateway` enabled. 
And, `gateway` used to depend on `http` and `utils` which is non-sensical. Both these things
are fixed now.

GnomedDev committed:
Only run MSRV job on the `current` branch (#2191)

GnomedDev committed:
Remove backports due to new MSRV policy (#2190)

kangalio committed:
Keep embeds in `EditInteractionResponse` by default (#2179)

This also adds a new `testing` crate meant to host small snippets to verify
correctness for all new pull requests.

kangalio committed:
Support interactions endpoint URL (#2186)

GnomedDev committed:
Fix clippy warnings (#2189)

kangalio committed:
Overhaul attachments (#2163)

kangalio committed:
Take `impl CacheHttp` instead of `&Context` in `argument_convert` (#2182)

kangalio committed:
Include new features in documentation (#2193)

Documentation had a hard coded list of features (two duplicated copies of them,
actually). The new `interactions_endpoint` was not among them, so that API was
completely absent from the docs.

This commit adds an internal `_docs` feature that enables all features that
should be documented. The doc generation now refers to `_docs` instead of
writing out the features

kangalio committed:
Replace `Message::is_private`; fix permission check for `Message::delete` (#2151)

kangalio committed:
Use `*Id` types everywhere (#2170)

kangalio committed:
Test JSON instead of serde tokens (#2200)

Removes serde_test dependency and cuts down on needless verbosity by testing
only what we really need to test. We need to verify *JSON* de-/serialization is
correct; where a lot of serde's extra metadata doesn't apply.

louboecker committed:
Return cached messages when a channel is deleted (#2194)

mkrasnitski committed:
Replace usage of `CacheAndHttp` in dispatch with `Context` (#2197)

mkrasnitski committed:
Bump MSRV to Rust 1.64 and raise edition to 2021 (#2199)

kangalio committed:
Fix `interactions_endpoint` documentation (#2201)

kangalio committed:
Remove lifetime from `CreateAttachment` (#2202)

kangalio committed:
Use stable Rust for docs CI instead of pinned nightly (#2206)

kangalio committed:
Actually include all types in `model::prelude` and make their docs inline  (#2204)

mkrasnitski committed:
Clean up the codebase related to the MSRV (#2208)

kangalio committed:
Use macro and `serde-derive` to shrink `audit_log/change.rs` (#2209)

kangalio committed:
Simplify serde (de)serialization implementations and remove `serde-value` dependency (#2211)

kangalio committed:
Clean up dependencies (#2212)

This:
- removes dependencies: `cfg-if`, `mime`, `tokio-test`
- makes dependencies optional: `fxhash`, `mime_guess`
- puts each dependency in `Cargo.toml` on a single line
- puts dependency attributes in consistent order (version, package/path, 
    default-features, features, optional)
- groups dependencies by category

kangalio committed:
Use nightly Rust for docs, to use `doc_auto_cfg` (#2207)

kangalio committed:
Put big model type fields behind `Box` (#2205)

This helps reduce the stack size of most types.

vaporoxx committed:
Replace `ClientBuilder`'s `Future` impl with `IntoFuture` (#2173)

kangalio committed:
Implement `Default` for some model types and shorten tests (#2215)

kangalio committed:
Simplify dispatch (#2203)

GnomedDev committed:
Use inline formatting (#2218)

jepcd committed:
Fix embeds being cleared when editing messages (#2216)

GnomedDev committed:
Implement `custom_ids` filter for `CollectorBuilder` (#2217)

kangalio committed:
Use `pub use X::*` in `builder/mod.rs` (#2225)

kangalio committed:
Make `CreateActionRow` an enum (#2227)

mkrasnitski committed:
Reduce extra clones in the `http` module (#2219)

arqunis committed:
Fix documentation building

kangalio committed:
Make `CreateInteractionResponse` an enum (#2228)

kangalio committed:
Add `CacheSettings::{time_to_live, cache_guilds, cache_channels, cache_users}` (#2210)

The idea is that even users on resource contrained devices can enable cache
feature, but just disable all the individual caching flags. What remains is the
cached fixed-size data: `CachedShardData` and more importantly `CurrentUser`.

kangalio committed:
Replace `CreateComponents` with `Vec<CreateActionRow>` (#2226)

It was just a dumb wrapper around `Vec<CreateActionRow>` anyways.

kangalio committed:
Implement new select menu types (#2224)

mkrasnitski committed:
Move `#[must_use]` to struct definition for `CreateGuildWelcomeChannel` (#2230)

kangalio committed:
Deduplicate CommandOption definition in builder (#2231)

arqunis committed:
Fix documentation building, again

kangalio committed:
Shorten type names (#2232)

Global find-and-replace:
- ApplicationCommand -> Command
- AutoModeration -> AutoMod
- GuildScheduledEvent -> ScheduledEvent
- MessageComponent -> Component

kangalio committed:
Fix `unused_mut` warning (#2242)

kangalio committed:
Remove `ButtonStyle` as a required argument of `CreateButton::new` (#2243)

kangalio committed:
Rename `ModalSubmit` to `Modal` (#2234)

`ModalInteraction` is just as clear as `ModalSubmitInteraction`

kangalio committed:
Shorten interaction method names (#2235)

`create_interaction_response -> `respond`
`create_followup_message` -> `respond_followup`
`delete_original_interaction_response` -> `delete_response`
`delete_followup_message` -> `delete_followup`
`edit_original_interaction_response` -> `edit_response`
`edit_followup_message` -> `edit_followup`
`get_interaction_response` -> `get_response`

kangalio committed:
Add `.button()` and `.select_menu()` builder convenience methods (#2233)

kangalioo commented:
I'd like to hold off updating the collectors until after #2240 is merged, because it's quite hard for me to work on the collector code as is

kangalio committed:
Flatten interaction

kangalio committed:
Fix function that parses errors from JSON responses (#2256)

kangalio committed:
Remove `create_autocomplete_response` (#2254)

#2228 changed `CreateAutocompleteResponse` to be a stand-alone response struct
into just the representation of the `data` response struct field. That's
because you are supposed to embed it in the `CreateInteractionResponse` enum
now. 

But I forgot that there exists `create_autocomplete_response`, a convenience
function for `create_response`, which took `CreateAutocompleteResponse`
directly. After #2228, `create_autocomplete_response` still compiled, but now
didn't send a proper response to discord but just its `data` field.

This commit removes `create_autocomplete_response` as a temporary fix. For now,
you should use `create_response` as for any other interaction response.

kangalio committed:
Use `foo/mod.rs` instead of `foo.rs` everywhere (#2253)

kangalio committed:
Avoid duplicating the `Timestamp` API (#2252)

kangalio committed:
Improve event docs (#2249)

- Update Discord docs links on event structs
- Add the required gateway intents to every event
- On `Cache`, add information which events are used for which cache resource (fixes #1601)

kangalio committed:
Remove `unstable_discord_api` flag from forum channels (#2247)

They are officially released now.

kangalio committed:
Remove `builder`->`utils` feature dependency (#2245)

kangalio committed:
Make absolute ratelimits a runtime config instead of a feature (#2246)

kangalio committed:
Add quick modal convenience feature (#2236)

jepcd committed:
Add support for multiple event handlers (#2257)

kangalio committed:
Make `label` on `Button` optional again (#2260)

mkrasnitski committed:
Bump MSRV to Rust 1.65 (#2267)

kangalio committed:
Rework `Framework` trait  (#2270)

GnomedDev committed:
Remove `nougat` dependency (#2268)

kangalio committed:
Add Reaction Remove Emoji event (#2273)

GnomedDev committed:
Avoid cloning entire cache fields by adding `ReadOnlyMapRef` (#2261)

kangalio committed:
Flatten `http` module (#2244)

kangalio committed:
Hide hasher implementation behind newtype (#2263)

kangalio committed:
Rewrite collectors (#2240)

arqunis committed:
Revert "Rewrite collectors (#2240)"

This reverts commit c259280c3b368c6d4c593596a3499f28038d6923.

I presumed it was Github being confused with which changes to show when
merging the pull request (#2240), but they were real, reverting a few
commits on `next` as a result.

kangalio commented:
Doubtful again if this is a good idea
